webhooks/cluster_test.go

```diff
@@ -69,8 +66,380 @@ func TestClusterDefaultTopologyVersion(t *testing.T) {
 		},
 	}
 	webhook := &Cluster{}
-	t.Run("for Cluster", customDefaultValidateTest(ctx, in, webhook))
-	g.Expect(webhook.Default(ctx, in)).To(Succeed())
+	t.Run("for Cluster", customDefaultValidateTest(ctx, c, webhook))
+	c.Default()
+
+	g.Expect(c.Spec.Topology.Version).To(HavePrefix("v"))
+}
+
+func TestClusterValidation(t *testing.T) {
+	// NOTE: ClusterTopology feature flag is disabled by default, thus preventing to set Cluster.Topologies.
+
+	tests := []struct {
+		name      string
+		in        *clusterv1.Cluster
+		old       *clusterv1.Cluster
+		expectErr bool
+	}{
+		{
+			name:      "should return error when cluster namespace and infrastructure ref namespace mismatch",
+			expectErr: true,
+			in: &clusterv1.Cluster{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: "foo",
+				},
+				Spec: clusterv1.ClusterSpec{
+					InfrastructureRef: &corev1.ObjectReference{
+						Namespace: "bar",
+					},
+					ControlPlaneRef: &corev1.ObjectReference{
+						Namespace: "foo",
+					},
+				},
+			},
+		},
+		{
+			name:      "should return error when cluster namespace and controlplane ref namespace mismatch",
+			expectErr: true,
+			in: &clusterv1.Cluster{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: "foo",
+				},
+				Spec: clusterv1.ClusterSpec{
+					InfrastructureRef: &corev1.ObjectReference{
+						Namespace: "foo",
+					},
+					ControlPlaneRef: &corev1.ObjectReference{
+						Namespace: "bar",
+					},
+				},
+			},
+		},
+		{
+			name:      "should succeed when namespaces match",
+			expectErr: false,
+			in: &clusterv1.Cluster{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: "foo",
+				},
+				Spec: clusterv1.ClusterSpec{
+					ControlPlaneRef: &corev1.ObjectReference{
+						Namespace: "foo",
+					},
+					InfrastructureRef: &corev1.ObjectReference{
+						Namespace: "foo",
+					},
+				},
+			},
+		},
+		{
+			name:      "fails if topology is set but feature flag is disabled",
+			expectErr: true,
+			in: &clusterv1.Cluster{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: "foo",
+				},
+				Spec: clusterv1.ClusterSpec{
+					ControlPlaneRef: &corev1.ObjectReference{
+						Namespace: "foo",
+					},
+					InfrastructureRef: &corev1.ObjectReference{
+						Namespace: "foo",
+					},
+					Topology: &clusterv1.Topology{},
+				},
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			g := NewWithT(t)
+			c := Cluster{}
```

Comment:
```suggestion
			webhook := Cluster{}
```
nit. I would suggest calling the var webhook, to avoid confusion with the clusterv1.Cluster struct

Reply:
Same below in ~ l.437